File: MergeSort/mergeSort.java
import java.util.Scanner;

public class mergeSort {

    // Merge two subarrays into arr[]
    public static void merge(int arr[], int si, int mid, int ei) {
        int temp[] = new int[ei - si + 1];
        int i = si;     // idx for 1st sorted part
        int j = mid + 1; // idx for 2nd sorted part
        int k = 0;      // index for temp

        while (i <= mid && j <= ei) {
            if (arr[i] < arr[j]) {
                temp[k++] = arr[i++];
            } else {
                temp[k++] = arr[j++];
            }
        }

        // Copy remaining elements of left half
        while (i <= mid) {
            temp[k++] = arr[i++];
        }

        // Copy remaining elements of right half
        while (j <= ei) {
            temp[k++] = arr[j++];
        }

        // Copy temp back to the original array
        for (k = 0, i = si; k < temp.length; k++, i++) {
            arr[i] = temp[k];
        }
    }

    // Main function that sorts arr[si..ei]
    public static void Mergesort(int arr[], int si, int ei) {
        if (si >= ei) {
            return;
        }

        int mid = (si + ei) / 2;
        Mergesort(arr, si, mid);
        Mergesort(arr, mid + 1, ei);
        merge(arr, si, mid, ei);
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter number of elements: ");
        int n = sc.nextInt();

        int arr[] = new int[n];
        System.out.println("Enter " + n + " elements:");
        for(int i = 0; i < n; i++) {
            arr[i] = sc.nextInt();
        }

        Mergesort(arr, 0, n - 1);

        System.out.println("Sorted array:");
        for(int i = 0; i < n; i++) {
            System.out.print(arr[i] + " ");
        }
    }
}

File: MergeSort.py
# Merge Sort Implementation in Python (with user input)

# Function to merge two sorted subarrays
def merge(arr, left, mid, right):
    # Create temp arrays
    L = arr[left:mid+1]
    R = arr[mid+1:right+1]

    # Indexes for temp arrays
    i = 0  # Left array index
    j = 0  # Right array index
    k = left  # Merged array index

    # Merge the temp arrays back into arr[left:right+1]
    while i < len(L) and j < len(R):
        if L[i] <= R[j]:
            arr[k] = L[i]
            i += 1
        else:
            arr[k] = R[j]
            j += 1
        k += 1

    # Copy any remaining elements of L[]
    while i < len(L):
        arr[k] = L[i]
        i += 1
        k += 1

    # Copy any remaining elements of R[]
    while j < len(R):
        arr[k] = R[j]
        j += 1
        k += 1


# Recursive function to sort the array
def merge_sort(arr, left, right):
    if left < right:
        # Find the middle point
        mid = (left + right) // 2

        # Sort first half
        merge_sort(arr, left, mid)

        # Sort second half
        merge_sort(arr, mid + 1, right)

        # Merge the sorted halves
        merge(arr, left, mid, right)


# --- Main Program ---
# Take number of elements from user
n = int(input("Enter the number of elements: "))

# Take array elements from user
arr = []
print("Enter the elements:")
for _ in range(n):
    arr.append(int(input()))

# Display original array
print("Original array:", arr)

# Call merge sort function
merge_sort(arr, 0, n - 1)

# Display sorted array
print("Sorted array:", arr)




#Enter the number of elements: 5
#Enter the elements:
#38
#27
#43
#3
#9
#Original array: [38, 27, 43, 3, 9]
#Sorted array: [3, 9, 27, 38, 43]

